dataplane/README.md

```diff
@@ -0,0 +1,176 @@
+1. saiserver: implements a "proto-ized" version of the [Switch Abstraction Interface API](https://github.com/opencomputeproject/SAI) and defines the forwarding pipeline
+1. apigen: code to generate the proto API for the saiserver and C++ code that calls the API
+1. dplanerc: gNMI reconciler that configures the dataplane based on OpenConfig schema
+1. standalone: shared library that a implements the C SAI API and container running only the dataplane
+
+There are two ways to run the dataplane. The first as part of Lemming, using the gNMI reconciliation. The second is a SAI implementation (e.g. in SONIC).
+
+## Components
+
+### forwarding
+
+The forwarding directory contains packet processing actions, tables, and ports. These are the building blocks of the dataplane: they can be combined to create the
+packet processing pipeline. The forwarding behavior is configured using the [gRPC API](../proto/forwarding/forwarding_service.proto).
+
+Generally, external users should not program the dataplane using this API directly. It is too flexible and low-table. Instead, the saiserver API should be used instead.
+
+Each port has a stack of input and output actions. The forwarding server processes an incoming packet using the input actions of the output.
+If after processing all the actions, there is an output port, then the server processes the output actions of the output port.
+
+The following sections describe the key parts of the forwarding server. The relevant packages and protobuf files contain more complete documentation.
+
+#### Actions
+
+Actions can modify the contents of the packet, output the packet etc...
+
+Some the important actions include:
+
+* Update: Set packet headers values
+* Transmit: Set the output port
+* Lookup: Lookup in a table
+
+Implementation: [fwdport](forwarding/fwdaction/actions)
+
+#### Ports
+
+Ports define how to read and write the packets. A port contains a list of input and output actions that define actions to take on a packet after reading and before writing.
+
+Note: If no actions are added to a port: all packets are dropped!
+
+Example ports types:
+
+1. Kernel: uses linux AF_PACKET
+2. TAP: uses ioctl and a file descriptor to create/read to TAP interface
+
+Implementation: [fwdport](forwarding/fwdport/ports)
+
+#### Tables
+
+Tables match against packet headers and perform actions. Each table has a default action list, and list of entries. An entry defines a set of field and values to match against
+and a list of the corresponding actions.
+
+>Note: After a packet is looked up in a table, the new actions are processed before any pending actions.
+
+Example table types:
+
+1. Prefix: longest matching prefix of header values
+2. Exact: exact match
+
+Implementation: [fwdport](forwarding/fwdtable)
+
+### saiserver
+
+The saiserver is a gRPC API generated from the [SAI API](https://github.com/opencomputeproject/SAI). It provides a simpler API for programming the dataplane.
```

Comment:
Is "simpler" the only reason? I'm thinking it also takes advantage of the fact that protobufs/gRPC is a widely-used API format that also allows for languages other than C++ to directly access it?